Report performance metrics docs followup (#139852)
elastic/kibana: dev_docs/tutorials/adding_performance_metrics.mdx

tags: ['kibana', 'onboarding', 'setup', 'performance', 'development', 'telemetry']
---

## Reporting performance events

### Simple performance events

Lets assume we intend to report the performance of a specific action called `APP_ACTION`.
In order to do so, we need to first measure the timing of that action.
Once we have the time measurment, we can use the `reportPerformanceMetricEvent` API to report it.

The most basic form of reporting would be:

```typescript
reportPerformanceMetricEvent(analytics, {
  eventName: APP_ACTION,     
  duration,                  // Duration in milliseconds
});
```

Once executed, the metric would be delivered to the `stack-telemetry` cluster, along side with the event's context.
The data is updated periodically, so you might have to wait up to 30 minutes to see your data in the index.

Once indexed, this metric will appear in `ebt-kibana` index. 
It we also be mapped into an additional index, dedicated to performance metrics.

> We recommend using the performance metrics indices, where you can expect the following structure:

Reviewer: I was assuming this is going to be default because `ebt-kibana-*-performance-metrics` is the only [index pattern](https://telemetry-v2-staging.elastic.dev/s/kibana-performance/app/management/kibana/indexPatterns) we have and default one, so considering that do we want to mention about this ?

  "_source": {
    "timestamp": "2022-08-31T11:29:58.275Z"
    "event_type": "performance_metric",                                 // All events share a common event type to simplify mapping
    "eventName": "dashboard_loaded",                                    // Event name as specified when reporting it
    "duration": 736,                                                    // Event duration as specified when reporting it
    "context": {                                                        // Context holds information identifying the deployment, version, application and page that generated the event
      "version": "8.5.0-SNAPSHOT",
      "applicationId": "dashboards",
      "page": "app",
      "entityId": "61c58ad0-3dd3-11e8-b2b9-5d5dc1715159",
      "branch": "main",
      "labels": {
        "journeyName": "flight_dashboard",
        ...
      }
      ...
    },
    ...
  },
}
```

### Performance events with breakdowns and metadata

Lets assume we are interested in benchmarking the performance of a more complex event `COMPLEX_APP_ACTION`, that is made up of two steps: 

>  - `STEP_A` where there are two possible flows `flow1` or `flow2` 

Reviewer: These steps and flows are a bit ambiguous imho, maybe we can provide example what can be those since in the example [here](https://github.com/elastic/kibana/pull/139852/files#diff-779dac43cbf7efbd7867f94bf1b4e300db3e2fabf654825752b6d377df76d5d9R93)it's also constant and it might be hard to understand what can be step and flow.

Author: I made the example a bit more specific

it's searchable and can be used to further breakdown event metrics.

- `Event name` We decided to report all metric events under a single event_type, so that we can avoid registering it multiple times. This requires us to add a special standard field for eventName.
- `Standardized fields` (duration) these fields represent common values that might be reported for performance. This can include duration for time based reports, status, heapSize, memorySize, etc. Standardized fields will be mapped as is to the index, so they can be visualized.
- `Free fields` (key1, value1, key2, value2, ..., key5, value5) A limited set of integer-numbered free key-value pairs that can be used to report additional use-case specific fields. Teams may use them to create their own internal sub-schema. For example, a team may choose to use key1 to always report the loading time of a certain component that is always present on the page. Free fields will be mapped, if present, to the index, so they can be visualized. If a free field becomes popular, it can be promoted to a standardized field, freeing up the free field for other use.

> - `Non-standard fields` (kibana_version, protocol ...) - additional fields may be added to an event by extending them to the `meta` object. However, those won't be automatically mapped to the index. Since the `meta` field is a Flattened Field, this means those fields will be searchable by default, but an additional field mapping would be required to visualize them.

Reviewer: I think it can be good to keep design considerations in the docs, maybe in separate section ? So we can communicate clearly with the intentions of having this schema. And we can get feedback based on these.

Author: We can refer to the GH issue where this was discussed maybe?
I don't want to unnecessarily duplicate this, especially as it doesn't effect the implementation.

Reviewer: I am not highly opinionated on that but I'd prefer to keep things in one place (and feeling like docs is better than issues in that regards) people usually don't go through all the links and read them. But it is my opinion feel free to resolve either way ☺️

> ### How to choose and measure events

Reviewer: Love all the new suggestions thanks for improving ❤️

>   Make sure that the event is clearly defined and consistent.

Reviewer: What do we really mean by consistent?

>   Consider the start point and end point of the measurment and what happens between those points.

Reviewer: ```suggestion
  Consider the start point and end point of the measurement and what happens between those points.
```

>   Try to balance event names specificity. Calling an event `load` is too generic, naving an event `tsvb-data-load` is too specific (instad the visualization 

Reviewer: ```suggestion
  Try to balance event names specificity. Calling an event `load` is too generic, calling an event `tsvb-data-load` is too specific (instead the visualization 
```
This is quite good recommendation  👍🏽

>   Make sure that measurment and reporting happens in a point of the code that is executed only once. 

Reviewer: ```suggestion
  Make sure that measurement and reporting happens in a point of the code that is executed only once. 
```
I'm now sure this is weird autocorrection thingy 🤣
